File: packages/sardis-sdk-js/src/telemetry.ts
/**
 * Sardis SDK Telemetry Module
 *
 * Provides automatic agent registration, heartbeat, event tracking,
 * and header injection for agent-to-dashboard sync.
 *
 * All telemetry operations are wrapped in try/catch — telemetry
 * never crashes or blocks the SDK.
 *
 * @packageDocumentation
 */

import type { TelemetryConfig, TelemetryEvent } from './types.js';

/** Max events to buffer before dropping oldest */
const MAX_BUFFER_SIZE = 1000;

/** Default heartbeat interval in seconds */
const DEFAULT_HEARTBEAT_INTERVAL = 60;

/** Default batch flush interval in seconds */
const DEFAULT_BATCH_INTERVAL = 10;

/** Default batch size (events per flush) */
const DEFAULT_BATCH_SIZE = 10;

/** Generate a random session ID */
function generateSessionId(): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = 'sess_';
  for (let i = 0; i < 24; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/** Generate a random agent ID if none provided */
function generateAgentId(): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let result = 'ag_';
  for (let i = 0; i < 16; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/**
 * Read an environment variable if available (Node.js or compatible runtimes).
 * Returns undefined in browser environments.
 */
function getEnv(key: string): string | undefined {
  try {
    if (typeof process !== 'undefined' && process.env) {
      return process.env[key];
    }
  } catch {
    // Browser or restricted environment
  }
  return undefined;
}

/**
 * SardisTelemetry handles agent auto-registration, heartbeat,
 * event batching, and correlation header injection.
 *
 * @example
 * ```typescript
 * const telemetry = new SardisTelemetry({
 *   agentId: 'my-agent',
 *   agentName: 'My Trading Agent',
 * });
 *
 * // Called internally by SardisClient
 * await telemetry.ensureRegistered(requestFn);
 * telemetry.startHeartbeat(requestFn);
 * telemetry.track('tool_call', { tool: 'get_balance' });
 * ```
 */
export class SardisTelemetry {
  private agentId: string;
  private agentName: string;
  private sessionId: string;
  private framework: string;
  private enabled: boolean;
  private heartbeatInterval: number;
  private batchInterval: number;
  private batchSize: number;

  private registered = false;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private flushTimer: ReturnType<typeof setInterval> | null = null;
  private buffer: TelemetryEvent[] = [];
  private shuttingDown = false;

  constructor(config?: TelemetryConfig) {
    const envEnabled = getEnv('SARDIS_TELEMETRY_ENABLED');
    this.enabled = config?.enabled ?? (envEnabled !== 'false');
    this.agentId = config?.agentId ?? getEnv('SARDIS_AGENT_ID') ?? generateAgentId();
    this.agentName = config?.agentName ?? getEnv('SARDIS_AGENT_NAME') ?? 'unnamed-agent';
    this.sessionId = generateSessionId();
    this.framework = config?.framework ?? '';

    const envHeartbeat = getEnv('SARDIS_HEARTBEAT_INTERVAL');
    this.heartbeatInterval = config?.heartbeatInterval
      ?? (envHeartbeat ? parseInt(envHeartbeat, 10) : DEFAULT_HEARTBEAT_INTERVAL);

    const envBatchInterval = getEnv('SARDIS_BATCH_INTERVAL');
    this.batchInterval = config?.batchInterval
      ?? (envBatchInterval ? parseInt(envBatchInterval, 10) : DEFAULT_BATCH_INTERVAL);

    const envBatchSize = getEnv('SARDIS_BATCH_SIZE');
    this.batchSize = config?.batchSize
      ?? (envBatchSize ? parseInt(envBatchSize, 10) : DEFAULT_BATCH_SIZE);
  }

  /**
   * Auto-register the agent with the server. Idempotent — safe to call multiple times.
   *
   * @param requestFn - The client's request function for making API calls
   */
  async ensureRegistered(
    requestFn: <T>(method: string, path: string, options?: { data?: unknown }) => Promise<T>
  ): Promise<void> {
    if (!this.enabled || this.registered) return;

    try {
      await requestFn('POST', '/api/v2/agents/auto-register', {
        data: {
          agent_id: this.agentId,
          name: this.agentName,
          session_id: this.sessionId,
          sdk_version: '1.0.0',
          framework: this.framework || undefined,
        },
      });
      this.registered = true;
    } catch {
      // Telemetry failure never propagates
    }
  }

  /**
   * Start periodic heartbeat to keep the agent marked as "online".
   * Also starts the event flush timer.
   *
   * @param requestFn - The client's request function for making API calls
   */
  startHeartbeat(
    requestFn: <T>(method: string, path: string, options?: { data?: unknown }) => Promise<T>
  ): void {
    if (!this.enabled) return;

    // Heartbeat with jitter to avoid thundering herd
    const sendHeartbeat = async () => {
      try {
        await requestFn('POST', '/api/v2/agents/heartbeat', {
          data: {
            agent_id: this.agentId,
            session_id: this.sessionId,
          },
        });
      } catch {
        // Silently ignore heartbeat failures
      }
    };

    // Jittered interval: interval * (0.8 + 0.4 * random())
    const jitteredMs = this.heartbeatInterval * 1000 * (0.8 + 0.4 * Math.random());
    this.heartbeatTimer = setInterval(sendHeartbeat, jitteredMs);

    // Unref the timer so it doesn't keep the process alive (Node.js)
    if (this.heartbeatTimer && typeof this.heartbeatTimer === 'object' && 'unref' in this.heartbeatTimer) {
      (this.heartbeatTimer as NodeJS.Timeout).unref();
    }

    // Start flush timer
    this.startFlushTimer(requestFn);
  }

  /**
   * Track an event. Events are buffered and flushed periodically.
   *
   * @param eventType - The type of event (e.g., 'tool_call', 'payment', 'error')
   * @param data - Arbitrary event data
   */
  track(eventType: string, data?: Record<string, unknown>): void {
    if (!this.enabled) return;

    try {
      const event: TelemetryEvent = {
        event_type: eventType,
        event_data: data ?? {},
        sdk_timestamp: new Date().toISOString(),
      };

      this.buffer.push(event);

      // Drop oldest events if buffer exceeds max
      while (this.buffer.length > MAX_BUFFER_SIZE) {
        this.buffer.shift();
      }
    } catch {
      // Never crash on tracking
    }
  }

  /**
   * Flush buffered events to the server.
   *
   * @param requestFn - The client's request function for making API calls
   */
  async flush(
    requestFn: <T>(method: string, path: string, options?: { data?: unknown }) => Promise<T>
  ): Promise<void> {
    if (!this.enabled || this.buffer.length === 0) return;

    try {
      // Take up to batchSize events
      const batch = this.buffer.splice(0, this.batchSize);

      await requestFn('POST', `/api/v2/agents/${this.agentId}/events/batch`, {
        data: {
          events: batch,
          session_id: this.sessionId,
        },
      });
    } catch {
      // Silently ignore flush failures — events are already removed from buffer
      // to avoid infinite retry loops
    }
  }

  /**
   * Stop heartbeat, flush remaining events, and clean up timers.
   *
   * @param requestFn - The client's request function for making API calls
   */
  async shutdown(
    requestFn: <T>(method: string, path: string, options?: { data?: unknown }) => Promise<T>
  ): Promise<void> {
    if (this.shuttingDown) return;
    this.shuttingDown = true;

    try {
      // Stop timers
      if (this.heartbeatTimer) {
        clearInterval(this.heartbeatTimer);
        this.heartbeatTimer = null;
      }
      if (this.flushTimer) {
        clearInterval(this.flushTimer);
        this.flushTimer = null;
      }

      // Final flush of remaining events
      while (this.buffer.length > 0) {
        await this.flush(requestFn);
      }
    } catch {
      // Never crash on shutdown
    }
  }

  /**
   * Returns correlation headers to merge into every API request.
   */
  getHeaders(): Record<string, string> {
    if (!this.enabled) return {};

    return {
      'X-Sardis-Agent-Id': this.agentId,
      'X-Sardis-Session-Id': this.sessionId,
    };
  }

  /**
   * Whether telemetry is enabled.
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * Start the periodic flush timer.
   */
  private startFlushTimer(
    requestFn: <T>(method: string, path: string, options?: { data?: unknown }) => Promise<T>
  ): void {
    this.flushTimer = setInterval(async () => {
      try {
        await this.flush(requestFn);
      } catch {
        // Silently ignore
      }
    }, this.batchInterval * 1000);

    // Unref so it doesn't keep the process alive
    if (this.flushTimer && typeof this.flushTimer === 'object' && 'unref' in this.flushTimer) {
      (this.flushTimer as NodeJS.Timeout).unref();
    }
  }
}

File: packages/sardis/src/sardis/telemetry.py
"""
Sardis SDK Telemetry — auto-registration, heartbeat, and event batching.

Never blocks, never crashes. All operations are best-effort with try/except.
Thread-safe for sync usage, asyncio-safe for async usage.
"""
from __future__ import annotations

import atexit
import logging
import os
import random
import threading
import time
import uuid
from collections import deque
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    import httpx

logger = logging.getLogger("sardis_sdk.telemetry")


@dataclass
class TelemetryConfig:
    """Configuration for SDK telemetry.

    Telemetry is OPT-IN: it is OFF by default and only activates when the
    consumer explicitly opts in (``SARDIS_TELEMETRY_ENABLED=true`` or
    ``TelemetryConfig(enabled=True)``). All fields can be overridden via
    environment variables.
    """

    enabled: bool = False
    agent_id: str | None = None
    agent_name: str = "unnamed-agent"
    framework: str | None = None
    heartbeat_interval: int = 60
    batch_size: int = 10
    batch_interval: int = 10

    @classmethod
    def from_env(cls, **overrides: Any) -> TelemetryConfig:
        """Build config from environment variables, with keyword overrides taking precedence."""
        # Opt-in: default OFF. Only the explicit truthy values turn it on.
        enabled_str = os.environ.get("SARDIS_TELEMETRY_ENABLED", "false")
        enabled = enabled_str.lower() in ("true", "1", "yes", "on")

        agent_id = os.environ.get("SARDIS_AGENT_ID")
        agent_name = os.environ.get("SARDIS_AGENT_NAME", "unnamed-agent")
        heartbeat_interval = int(os.environ.get("SARDIS_HEARTBEAT_INTERVAL", "60"))
        batch_size = int(os.environ.get("SARDIS_BATCH_SIZE", "10"))
        batch_interval = int(os.environ.get("SARDIS_BATCH_INTERVAL", "10"))

        cfg = cls(
            enabled=enabled,
            agent_id=agent_id,
            agent_name=agent_name,
            heartbeat_interval=heartbeat_interval,
            batch_size=batch_size,
            batch_interval=batch_interval,
        )
        for k, v in overrides.items():
            if v is not None and hasattr(cfg, k):
                setattr(cfg, k, v)
        return cfg


@dataclass
class TelemetryEvent:
    """A single telemetry event queued for batching."""

    event_type: str
    data: dict[str, Any] = field(default_factory=dict)
    timestamp: float = field(default_factory=time.time)


class SardisTelemetry:
    """SDK telemetry: auto-registration, heartbeat, and event batching.

    Designed to be completely non-blocking and crash-proof.
    All public methods swallow exceptions and log at debug level.
    """

    MAX_QUEUE_SIZE = 1000

    def __init__(
        self,
        config: TelemetryConfig,
        base_url: str,
        api_key: str,
        sdk_version: str,
    ) -> None:
        self._config = config
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._sdk_version = sdk_version

        self._session_id = str(uuid.uuid4())
        self._agent_id: str | None = config.agent_id
        self._registered = False

        # Event queue (thread-safe via lock)
        self._queue: deque[TelemetryEvent] = deque(maxlen=self.MAX_QUEUE_SIZE)
        self._lock = threading.Lock()

        # Background threads
        self._heartbeat_timer: threading.Timer | None = None
        self._flush_timer: threading.Timer | None = None
        self._shutdown_flag = threading.Event()

        # Lazy httpx client for telemetry calls
        self._http: httpx.Client | None = None

    # ------------------------------------------------------------------
    # Internal HTTP
    # ------------------------------------------------------------------

    def _get_http(self) -> httpx.Client:
        if self._http is None or self._http.is_closed:
            import httpx

            self._http = httpx.Client(
                base_url=self._base_url,
                timeout=httpx.Timeout(5.0),
                headers={
                    "X-API-Key": self._api_key,
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
        return self._http

    def _post(self, path: str, json_body: dict[str, Any]) -> dict[str, Any] | None:
        """Fire a POST request. Returns response JSON or None on failure."""
        try:
            resp = self._get_http().post(path, json=json_body)
            if resp.status_code < 400:
                return resp.json()
            logger.debug("Telemetry POST %s returned %d", path, resp.status_code)
        except Exception:
            logger.debug("Telemetry POST %s failed", path, exc_info=True)
        return None

    # ------------------------------------------------------------------
    # Layer 2: Registration + Heartbeat
    # ------------------------------------------------------------------

    def ensure_registered(self) -> None:
        """Register agent with the server (idempotent). Never raises."""
        if not self._config.enabled or self._registered:
            return
        try:
            payload: dict[str, Any] = {
                "name": self._config.agent_name,
                "session_id": self._session_id,
                "sdk_version": self._sdk_version,
            }
            if self._agent_id:
                payload["agent_id"] = self._agent_id
            if self._config.framework:
                payload["framework"] = self._config.framework

            result = self._post("/api/v2/agents/auto-register", payload)
            if result:
                self._agent_id = result.get("agent_id", self._agent_id)
                self._registered = True
                logger.debug(
                    "Agent registered: id=%s, created=%s",
                    self._agent_id,
                    result.get("created", False),
                )
        except Exception:
            logger.debug("ensure_registered failed", exc_info=True)

    def start_heartbeat(self) -> None:
        """Start the background heartbeat thread. Never raises."""
        if not self._config.enabled:
            return
        try:
            self._schedule_heartbeat()
            atexit.register(self.shutdown)
        except Exception:
            logger.debug("start_heartbeat failed", exc_info=True)

    def _schedule_heartbeat(self) -> None:
        if self._shutdown_flag.is_set():
            return
        # Jitter: interval * (0.8 + 0.4 * random()) to avoid thundering herd
        jittered = self._config.heartbeat_interval * (0.8 + 0.4 * random.random())
        self._heartbeat_timer = threading.Timer(jittered, self._heartbeat_tick)
        self._heartbeat_timer.daemon = True
        self._heartbeat_timer.start()

    def _heartbeat_tick(self) -> None:
        if self._shutdown_flag.is_set():
            return
        try:
            payload: dict[str, Any] = {"session_id": self._session_id}
            if self._agent_id:
                payload["agent_id"] = self._agent_id
            self._post("/api/v2/agents/heartbeat", payload)
        except Exception:
            logger.debug("heartbeat tick failed", exc_info=True)
        self._schedule_heartbeat()

    # ------------------------------------------------------------------
    # Layer 3: Event Batching
    # ------------------------------------------------------------------

    def track(self, event_type: str, data: dict[str, Any] | None = None) -> None:
        """Queue a telemetry event. Never blocks, never raises."""
        if not self._config.enabled:
            return
        try:
            event = TelemetryEvent(event_type=event_type, data=data or {})
            with self._lock:
                self._queue.append(event)
                queue_len = len(self._queue)

            # Auto-flush if batch size reached
            if queue_len >= self._config.batch_size:
                self._do_flush()
        except Exception:
            logger.debug("track failed", exc_info=True)

    def start_flush_timer(self) -> None:
        """Start periodic flush timer. Never raises."""
        if not self._config.enabled:
            return
        try:
            self._schedule_flush()
        except Exception:
            logger.debug("start_flush_timer failed", exc_info=True)

    def _schedule_flush(self) -> None:
        if self._shutdown_flag.is_set():
            return
        self._flush_timer = threading.Timer(
            float(self._config.batch_interval), self._flush_tick
        )
        self._flush_timer.daemon = True
        self._flush_timer.start()

    def _flush_tick(self) -> None:
        if self._shutdown_flag.is_set():
            return
        self._do_flush()
        self._schedule_flush()

    def flush(self) -> None:
        """Flush queued events immediately. Never raises."""
        try:
            self._do_flush()
        except Exception:
            logger.debug("flush failed", exc_info=True)

    def _do_flush(self) -> None:
        """Drain the queue and POST events batch to the server."""
        if not self._agent_id:
            return

        with self._lock:
            if not self._queue:
                return
            batch = list(self._queue)
            self._queue.clear()

        events_payload = [
            {
                "event_type": e.event_type,
                "data": e.data,
                "sdk_timestamp": e.timestamp,
            }
            for e in batch
        ]

        self._post(
            f"/api/v2/agents/{self._agent_id}/events/batch",
            {"events": events_payload, "session_id": self._session_id},
        )

    # ------------------------------------------------------------------
    # Headers
    # ------------------------------------------------------------------

    def get_headers(self) -> dict[str, str]:
        """Return telemetry headers to merge into API requests."""
        headers: dict[str, str] = {}
        if self._agent_id:
            headers["X-Sardis-Agent-Id"] = self._agent_id
        headers["X-Sardis-Session-Id"] = self._session_id
        return headers

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def shutdown(self) -> None:
        """Stop heartbeat, flush remaining events, close HTTP client. Never raises."""
        try:
            self._shutdown_flag.set()

            if self._heartbeat_timer:
                self._heartbeat_timer.cancel()
                self._heartbeat_timer = None

            if self._flush_timer:
                self._flush_timer.cancel()
                self._flush_timer = None

            self._do_flush()

            if self._http and not self._http.is_closed:
                self._http.close()
                self._http = None
        except Exception:
            logger.debug("shutdown failed", exc_info=True)


# ---------------------------------------------------------------------------
# Async variant
# ---------------------------------------------------------------------------


class AsyncSardisTelemetry:
    """Async telemetry using asyncio tasks instead of threads.

    Same API surface as SardisTelemetry but uses async HTTP and asyncio timers.
    """

    MAX_QUEUE_SIZE = 1000

    def __init__(
        self,
        config: TelemetryConfig,
        base_url: str,
        api_key: str,
        sdk_version: str,
    ) -> None:
        self._config = config
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._sdk_version = sdk_version

        self._session_id = str(uuid.uuid4())
        self._agent_id: str | None = config.agent_id
        self._registered = False

        self._queue: deque[TelemetryEvent] = deque(maxlen=self.MAX_QUEUE_SIZE)
        self._shutdown_flag = False

        self._heartbeat_task: Any = None
        self._flush_task: Any = None
        self._http: httpx.AsyncClient | None = None

    async def _get_http(self) -> httpx.AsyncClient:
        if self._http is None or self._http.is_closed:
            import httpx

            self._http = httpx.AsyncClient(
                base_url=self._base_url,
                timeout=httpx.Timeout(5.0),
                headers={
                    "X-API-Key": self._api_key,
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
        return self._http

    async def _post(self, path: str, json_body: dict[str, Any]) -> dict[str, Any] | None:
        try:
            client = await self._get_http()
            resp = await client.post(path, json=json_body)
            if resp.status_code < 400:
                return resp.json()
            logger.debug("Async telemetry POST %s returned %d", path, resp.status_code)
        except Exception:
            logger.debug("Async telemetry POST %s failed", path, exc_info=True)
        return None

    async def ensure_registered(self) -> None:
        if not self._config.enabled or self._registered:
            return
        try:
            payload: dict[str, Any] = {
                "name": self._config.agent_name,
                "session_id": self._session_id,
                "sdk_version": self._sdk_version,
            }
            if self._agent_id:
                payload["agent_id"] = self._agent_id
            if self._config.framework:
                payload["framework"] = self._config.framework

            result = await self._post("/api/v2/agents/auto-register", payload)
            if result:
                self._agent_id = result.get("agent_id", self._agent_id)
                self._registered = True
                logger.debug(
                    "Agent registered (async): id=%s, created=%s",
                    self._agent_id,
                    result.get("created", False),
                )
        except Exception:
            logger.debug("ensure_registered (async) failed", exc_info=True)

    async def start_heartbeat(self) -> None:
        if not self._config.enabled:
            return
        try:
            import asyncio

            self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())
        except Exception:
            logger.debug("start_heartbeat (async) failed", exc_info=True)

    async def _heartbeat_loop(self) -> None:
        import asyncio

        while not self._shutdown_flag:
            jittered = self._config.heartbeat_interval * (0.8 + 0.4 * random.random())
            await asyncio.sleep(jittered)
            if self._shutdown_flag:
                break
            try:
                payload: dict[str, Any] = {"session_id": self._session_id}
                if self._agent_id:
                    payload["agent_id"] = self._agent_id
                await self._post("/api/v2/agents/heartbeat", payload)
            except Exception:
                logger.debug("heartbeat tick (async) failed", exc_info=True)

    def track(self, event_type: str, data: dict[str, Any] | None = None) -> None:
        """Queue event (sync-safe, can be called from any context)."""
        if not self._config.enabled:
            return
        try:
            event = TelemetryEvent(event_type=event_type, data=data or {})
            self._queue.append(event)
        except Exception:
            logger.debug("track (async) failed", exc_info=True)

    async def start_flush_timer(self) -> None:
        if not self._config.enabled:
            return
        try:
            import asyncio

            self._flush_task = asyncio.create_task(self._flush_loop())
        except Exception:
            logger.debug("start_flush_timer (async) failed", exc_info=True)

    async def _flush_loop(self) -> None:
        import asyncio

        while not self._shutdown_flag:
            await asyncio.sleep(float(self._config.batch_interval))
            if self._shutdown_flag:
                break
            await self._do_flush()

    async def flush(self) -> None:
        try:
            await self._do_flush()
        except Exception:
            logger.debug("flush (async) failed", exc_info=True)

    async def _do_flush(self) -> None:
        if not self._agent_id or not self._queue:
            return

        batch = list(self._queue)
        self._queue.clear()

        events_payload = [
            {
                "event_type": e.event_type,
                "data": e.data,
                "sdk_timestamp": e.timestamp,
            }
            for e in batch
        ]

        await self._post(
            f"/api/v2/agents/{self._agent_id}/events/batch",
            {"events": events_payload, "session_id": self._session_id},
        )

    def get_headers(self) -> dict[str, str]:
        headers: dict[str, str] = {}
        if self._agent_id:
            headers["X-Sardis-Agent-Id"] = self._agent_id
        headers["X-Sardis-Session-Id"] = self._session_id
        return headers

    async def shutdown(self) -> None:
        try:
            self._shutdown_flag = True

            if self._heartbeat_task and not self._heartbeat_task.done():
                self._heartbeat_task.cancel()
                self._heartbeat_task = None

            if self._flush_task and not self._flush_task.done():
                self._flush_task.cancel()
                self._flush_task = None

            await self._do_flush()

            if self._http and not self._http.is_closed:
                await self._http.aclose()
                self._http = None
        except Exception:
            logger.debug("shutdown (async) failed", exc_info=True)
